Feature/ci add cppcheck common

rmelotte opened the pull request:
Add a CI job to run cppcheck on common, save the results as an artifact and output a summary.

Only issues of severity 'warning' and 'error' are enforced for now.

Fix the issues of severity 'warning' and 'error', and some of severity style.

rmelotte committed:
tools: docker: add a script to run cppcheck and summarize results

Add a new script, cppcheck.sh, to run cppcheck and output the number
of issues per severity. The results are both saved in a file and
output to stdout.

The script can be used by giving any number of paths to run cppcheck
on. It exits with a status of 1 if any issue (of a severity different
than information) is found.

Issues of severity information are kept in the results so that we can
track issues with cppcheck configuration, but they are not output nor
influencing the return status.

New cppcheck issues will appear on the master branch at the same time
we fix issues on a side branch. It it's easier to first enforce all
cppcheck issues of severity warning and error throughout the
code (because there are less of those) than to try to fix all
severity types at once, even on parts of the code.

Return with an exit code of 0 if no issue of severity warning or error
is found.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
.gitlab-ci.yml: run and enforce cppcheck

Run cppcheck in CI, currently on the common folder only (other folders
will be added once they no longer have issues with the severity levels
we enforce).

The results are kept as an artifact to allow developers to check the
results in detail.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
framework: fix cppcheck error (shared_ptr dealloc)

When an explicit call to shared_ptr is made when calling
std::vector.push_back(), cppcheck believes the ptr is deallocated, and
should not be returned.

While it's a bug in cppcheck, the additional call to shared_ptr() is
not necessary, as it will create one more temporary ptr which is
immediately copied into the container.

Remove the call to shared_ptr to avoid the useless temporary.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
framework: fix cppcheck warning (unsigned comparison less than 0)

m_*_idx__ variables are always of type size_t, so no need to check if
they are negative.

Turn '<= 0' in into '== 0'.

Fix 27 ccpcheck issues with severity style.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
framework: tlvf: make single argument constructor explicit (cppcheck)

Constructors with only one mandatory argument can be used as
converting constructors when explicitly called, but also implicitly.

Since the constructor here is not meant to be used as a converting
constructor, make it explicit to avoid any potential implicit,
unwanted conversion.

Fix 368 cppcheck issues with severity style.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
common: fix call to virtual function in destructor (cppcheck)

Every bwl makes a call to a virtual function in its
destructor. cppcheck reports it as a warning because when called from
a destructor/constructor, only the function from the same class will
be called, not the derived ones.

Fix it for now by adding the explicitly qualified ID of the function,
to make it clear that we will only call the class's function and not
any derived one.

Another way to fix it would have been to make all the derived classes
final to make it even more clear that there is no derived class to
call the function from. This is however not possible, as
sta_wlan_hal_dummy inherits from base_wlan_hal_dummy which inherits
from base_wlan_hal, so we can't make the function final in
base_clan_hal_dummy (sta_wlan_hal overrides it).

In the future and depending on the implementation of the detach()
function, we might also want to solve it in a different way: split
detach() by using a private function m_detach() that every derived
class implement and that is called from both the destructor and the
virtual detach() function.

Fixes 5 cppcheck issues of severity warning.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
common: fix usage of virtual function in constructor/destructor

In beerocks_thread_base.cpp, the destructor was calling the virtual
function before_stop() (through the stop function). It looks like the
original intention was to call derived implementations, because the
implementation in the base class is empty.

Since classes are destructed by calling the most derived destructors
first, the overridden implementation of before_stop() is no longer
available when the base destructor is called!

Solve it by reproducing the behavior of the stop() function in the
base class destructor, and by adding a destructor to the derived
classes that calls its own before_stop().

common/beerocks/btl/btl.cpp: call the virtual function
set_select_timeout() with it's explicitly qualified ID since it's
called from the constructor.

common/beerocks/bcl/source/beerocks_socket_thread.cpp: call the
virtual function set_select_timeout() with it's explicitly qualified
ID since it's called from the constructor.

Fix 3 cppcheck issues of severity warning.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
framework: fix duplInheritedMember (cppcheck)

Class InterfaceConfigurationIndicationMessage previously inherited
from class InterfaceConfigurationRequestMessage because it's
essentially the same, except for the kTopicPrefix value.
However, since kTopicPrefix is static, the derived class could not
change its value without hiding it.

Move the common parts to a new base class:
InterfaceConfigurationMessage, and make the 2 classes inherit from it.

Fix 1 cppcheck issue of severity warning.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

tomereli commented on tools/docker/static-analysis/cppcheck.sh:
```diff
@@ -0,0 +1,89 @@
+
+. "${rootdir}/tools/docker/functions.sh"
+
+OUTPUT_FILE="$rootdir/cppcheck_results.txt"
```
why not add the output in an 'out' directory in this folder? (nitpick)

rmelotte replied:
We could, although we don't have such a folder for anything else at the moment.
In our case the scrupt is mostly meant to be used in a docker container though, because the results highly depend on the `cppcheck` version.

tomereli commented on tools/docker/static-analysis/cppcheck.sh:
```diff
@@ -0,0 +1,89 @@
+    msg="$2"
+    case "$severity" in
+        error)
+            printf '\033[1;31m%s\033[0m\n' "$msg"
```
Why not include `functions.sh` and use `err` `info` and friends?

rmelotte replied:
Because there is 1 color per severity type. `functions.sh` only has `err` and `success`. I didn't add them to `functions.sh` because I don't see any other use for them at the moment.

tomereli commented on tools/docker/static-analysis/cppcheck.sh:
```diff
@@ -0,0 +1,89 @@
+}
+
+main() {
+    if ! command -v cppcheck > /dev/null ; then
```
Please add cppcheck to the README.md in the prerequisites step.

rmelotte replied:
This comment makes me realize I misunderstood your similar comment on a previous PR.

`cppcheck` is not a pre-requisite to build prplMesh, so to me it doesn't belong to that list (the same is true for `uuidgen`, it's only needed if you want to build an ipk using the docker script).
Also, as the results really depend on the cppcheck version, I don't see much use in running it with the Ubuntu version.

What we could do if there is interest for it, is to add a section on how to start a docker container running cppcheck the same way the CI does.

tomereli replied:
This would be great - then people tasked with fixing cppcheck issues can use it instead of relying on CI.

rmelotte replied:
Added to the wiki: https://github.com/prplfoundation/prplMesh/wiki/Testing-in-CI

tomereli commented on .gitlab-ci.yml:
```diff
@@ -5,6 +5,16 @@
+  image: prplfoundationinc/prplmesh-builder:alpine-3.11.3
+  stage: build
+  script:
+    - tools/docker/static-analysis/cppcheck.sh common
```
Since eventually, we will run it on all folders - why not just make cppcheck.sh accept no argument and internally only run the check on common for now?

tomereli replied:
Also, why do we run it only on common? why not run it on all folders and be done with it? Can you explain the reason in the commit message?

rmelotte replied:
I wanted to keep the usage of `cppcheck.sh` flexible, to be able to call it on any folder namely (you can even give options to `cppcheck.sh`, they will be passed to cppcheck).

In CI we can only run it on `common` for now because the other folders still have many issues of severity 'warning' and 'error' to solve first. The job is made to fail if any of those issues remain.

tomereli replied:
So you can add an optional parameter which is the directory to check. If none given, default should be to check all.

tomereli commented on .gitlab-ci.yml:
```diff
@@ -5,6 +5,16 @@
     - docker
 
+cppcheck:
+  image: prplfoundationinc/prplmesh-builder:alpine-3.11.3
```
Why are we using the apline builder for cppcheck? This should be mentioned in the commit message.

rmelotte replied:
Done in a fixup (we use alpine because the version in Ubuntu 18.04 is too old).

tomereli commented on common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_1905_vs.h:
```diff
@@ -30,7 +30,7 @@ class tlvVsClientAssociationEvent : public BaseClass
     public:
         tlvVsClientAssociationEvent(uint8_t* buff, size_t buff_len, bool parse = false);
-        tlvVsClientAssociationEvent(std::shared_ptr<BaseClass> base, bool parse = false);
+        explicit tlvVsClientAssociationEvent(std::shared_ptr<BaseClass> base, bool parse = false);
```
The commit message is not very clear.

rmelotte replied:
Ok, I will explain the issue in more detail when I rebase.

tomereli reviewed (approved):
Mostly nitpicks, so approving.

rmelotte commented:
Updated to use a compilation database instead, and rebased on master.
Because we don't want to display the issues for folders other than common (until the issues for other folders are fixed), there remain a lot of " * [unmatchedSuppression]" which we would have to ignore.

arnout commented on tools/docker/static-analysis/cppcheck.sh:
```diff
@@ -0,0 +1,96 @@
+
+OUTPUT_FILE="$rootdir/cppcheck_results.txt"
+INCLUDES="\
+-I$rootdir/framework/platform/bpl/include \
```
(nitpick) I'd prefer the contents of the variable to be indented.

arnout commented on tools/docker/static-analysis/cppcheck.sh:
```diff
@@ -0,0 +1,96 @@
+
+run_cppcheck() {
+    # --force is needed to check more than 12 ifdefs
+    cppcheck --template='{file}:{line}:{column}: {severity}: {message} [{id}]\n{code}' --force --error-exitcode=1 --enable=warning,style,information -q $INCLUDES "$@" 2>&1 | tee "$OUTPUT_FILE"
```
(nitpick) Line is too long.

arnout commented on tools/docker/static-analysis/cppcheck.sh:
```diff
@@ -0,0 +1,96 @@
+            case "$s" in
+                # report the following severities in exit code:
+                error|warning)
+                    status=$((status + 1))
```
Note that if we have exactly 256 errors+warnings, this will overflow back to 0. Not important enough to fix IMO though.

rmelotte replied:
Actually the exact value of the exit code is not used, only the fact that it's non-zero, so I'll just change it to `status=1`.

arnout commented on tools/docker/static-analysis/cppcheck.sh:
```diff
@@ -0,0 +1,96 @@
+                # report the following severities in exit code:
+                error|warning)
+                    status=$((status + 1))
+                    ;;
```
I think it would be good to also show these on the output, so you don't have to hunt down the artefacts to fix issues. Something as simple as the grep above without the -c should be sufficient.

However, that should probably be done before the summary...

Ideal output would be:
```
WARNING
WARNING
WARNING
ERROR
ERROR
ERROR
3 issues with severity error
3 issues with severity warning
0 issues with severity style
```

i.e., first all warnings, then all errors, then the summary. This makes sure the errors are clearly visible at the end.

(nitpick)

rmelotte replied:
You mean you want the issues to be sorted by severity level?
In that case I can wait until cppcheck is done, then output them sorted.
If I keep the current behavior and add the warnings/errors at the end, they will be output 2 times...

arnout replied:
Never mind, it was absolutely a "nice-to-have". Please merge ASAP.

arnout commented on .gitlab-ci.yml:
```diff
@@ -5,6 +5,28 @@
   tags:
     - docker
 
+.with-compilation-database:
```
This compilation database has a little disadvantage: it limits the cppcheck to the "linux" headers and sources. So only the dummy bwl and bpl are checked.

This can be fixed later by adding different variants (but that's more work because we need stubs for the different libraries that are used by bwl/bpl).

rmelotte replied:
In that case, maybe it would be better to keep using the hardcoded list of includes in the meantime? What do you think?

rmelotte replied:
I dropped it and used the static list of includes for now.

arnout commented on tools/docker/builder/Dockerfile:
```diff
@@ -15,6 +15,7 @@ RUN apt-get update && apt-get install -y \
     binutils \
     clang-format \
     cmake \
+    cppcheck \
```
We actually don't use it in this image anymore, so I guess this commit is not needed.

Do we have a Dockerfile for the Alpine image?

rmelotte replied:
Indeed it's a mistake. The alpine image is already there and already contains cppcheck (`tools/docker/builder/alpine/Dockerfile`).

arnout commented on common/beerocks/bcl/source/beerocks_thread_base.cpp:
```diff
@@ -38,7 +38,7 @@ void thread_base::join()
 {
     should_stop = true;
-    before_stop();
+    thread_base::before_stop();
```
I don't think this is correct... This is declared empty in thread_base, so if the intention was to call the one from thread_base, it would just have been left out.

But indeed, the code is in fact broken, because when called from the destructor, the derived class's before_stop is no longer available!

So, I think the proper fix is to:
- duplicate the code below to the destructor instead of calling stop();
- add a destructor to async_work_queue that calls its own before_stop().

This is not a nitpick!

tomereli replied:
Is this resolved?

rmelotte replied:
Yes, done. Nice catch BTW!

arnout commented on framework/transport/ieee1905_transport/include/mapf/transport/ieee1905_transport_messages.h:
```diff
@@ -304,10 +302,6 @@ class InterfaceConfigurationRequestMessage : public Message {
 
-    virtual const std::string topic_prefix() const { return kTopicPrefix; }
-
-    virtual const std::string topic() const { return topic_prefix(); }
```
Can't this one stay in this class?

rmelotte replied:
Indeed. Done.

arnout reviewed (approved):
Mostly nitpicks, except for one bug that you introduced. But you can fix that before merging.

This one will require full rebase before merging, since new cppcheck failures may have been introduced.

arnout commented on common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_1905_vs.h:
```diff
@@ -30,7 +30,7 @@ class tlvVsClientAssociationEvent : public BaseClass
     public:
         tlvVsClientAssociationEvent(uint8_t* buff, size_t buff_len, bool parse = false);
-        tlvVsClientAssociationEvent(std::shared_ptr<BaseClass> base, bool parse = false);
+        explicit tlvVsClientAssociationEvent(std::shared_ptr<BaseClass> base, bool parse = false);
```
If everyone new to the project asks this question, maybe there's something wrong with the answer :-)

The answer is:

-----

By putting the AutoGenerated source in revision control, it is easier for someone to browse the code after a simple checkout or even online, without building it. Even for review it is sometimes easier, e.g. when tlvf is changed it's more relevant to review the generated code than the python code.

And to make sure the AutoGenerated code in git is always consistent with the yaml source, we have CI that verifies it.

-----

I also wouldn't have chosen to put the generated code in git, but I don't feel so strongly about it that I'm going to fight the decision. I believe @vitalybu is the only real proponent.

tomereli replied:
What Arnout said is the answer and I fully agree with it. If you want we can put it up for a vote again..

vitalybu replied:
I actually was against it at first as well, but the reasons Arnout mentioned eventually convinced me.

arnout replied:
Hang on, so who actually wants the generated code in git? IIUC @tomereli, @vitalybu, @arnout and @fvbogaert are all saying that they would prefer the generated code to *not* be in git, but that they see the arguments and are not opposed to it either. So who is left as a proponent? @morantr ? @rmelotte ?

Just to be clear, some advantages of *not* having the generated code in git:

- There is no risk that the generated code is inconsistent with the yaml files.
- It makes it clearer that this is generated code that shouldn't be modified.
- It removes some commits ("update autogenerated") and makes others easier to review.
- We can remove the check from CI.

rmelotte replied:
I don't like having the generated code in the repository either.
However in this case, the way to generate it is nothing "standard" or easy to read or check for errors, so I also agree for reviews it can help to have it included.

tomereli replied:
@arnout - The main disadvantage of removing the autogenerated files from git is reviewing changes to tlvf.py. Unfortunately, tlvf.py is not a very readable python module, so I think that a followup commit which shows the change is an advantage, not a disadvantage.
I assume @morantr and @LiorAmram would agree on this.

RanRegev replied:
M2c for this discussion -
Some PDFs which are the output of a complicated 'compilation' process are kept in source control. E.g. https://github.com/wg21il/Papers/tree/P2022/master/P2022/generated
These files are the "final product" and are the subject for review by humans.

Intermediate code, of all kinds, that is meant to be consumed mainly by other code, IMO should be kept locally, not in a source control, the reasons @arnout popinted.

The main issue that I see, which legitimate the use of git for generated code is human review.
So maybe we are using an incorrect tool to solve the problem.
Maybe we should address the review-auto-generated-code problem differently.
E.g. Supply a really easy-to-use generator.

tomereli replied:
OK @RanRegev, which easy-to-use generator do you have in mind?

RanRegev replied:
@tomereli, 
1. I suggested that sometimes we choose the wrong tool to solve a real problem.
If we agree that the main reason to save auto generated files in git is to enable humans review them - then maybe it is the wrong tool.
2. As for your question:
a. the [documentation](https://github.com/prplfoundation/prplMesh/blob/master/framework/tlvf/README.md) does not help. It says that there are 2 argument needed, but it calls for at least 3. I spent 10 minutes on it and I still don't know how to use tlvf.py and compile even a single file. Yes, maybe I'm a slow catcher and hard to understand, but 10 minutes are too much.
b. I understand the that _final_ requirement is that this script runs via the build system: bring other headers with it, compile many files together, have debug flags, and more. But this use of the script doesn't mean that it has to be complicated from the first place. For simple review it will be enough to have `tlvf.py <src-yaml>` - and that it is. No configuration file, not even an output file (stdout is great).
On top of this simplicity we can add as much complexity as we want. 
Having broad ranges of opportunities to compile tlv may, maybe, will remove the need store products in git.

tomereli replied:
> @tomereli,
> 
> 1. I suggested that sometimes we choose the wrong tool to solve a real problem.
>    If we agree that the main reason to save auto generated files in git is to enable humans review them - then maybe it is the wrong tool.

I still don't understand your point - what is the problem you are referring to and what is the correct tool in your opinion we should use?

> 2. As for your question:
>    a. the [documentation](https://github.com/prplfoundation/prplMesh/blob/master/framework/tlvf/README.md) does not help. It says that there are 2 argument needed, but it calls for at least 3. I spent 10 minutes on it and I still don't know how to use tlvf.py and compile even a single file. Yes, maybe I'm a slow catcher and hard to understand, but 10 minutes are too much.

The tlvf generates (not compiles) multiple files, there is no support to generate a single file. We didn't hear any complaints till now about TLVF from other developers. Perhaps if you struggle with TLVF than ask someone from your team to guide you, almost everyone in the community knows how to add a new class / structure to TLVF. Raising complaints or arguments that TLVF is hard to use is pointless - we have plans for TLVFv2 but its a huge effort and since TLVFv1 just works, there is no plan to replace it anytime soon. Its not like we had a wide array of tools to select from and we chose TLVF - there was none.

>    b. I understand the that _final_ requirement is that this script runs via the build system: bring other headers with it, compile many files together, have debug flags, and more. But this use of the script doesn't mean that it has to be complicated from the first place. For simple review it will be enough to have `tlvf.py <src-yaml>` - and that it is. No configuration file, not even an output file (stdout is great).
>    On top of this simplicity we can add as much complexity as we want.
>    Having broad ranges of opportunities to compile tlv may, maybe, will remove the need store products in git.

Even if this was supported by TLVF, it wouldn't help to review TLVF changes. IMO nothing we change in TLVF will make any difference - unless we trust TLVF as a complete black box, then we can just use the YAML files and ignore the autogenerated files altogether. We are actually pretty close to this now, TLVF supports almost everything we need.

Regarding your suggestion for having tlvf option which accepts a single yaml file - this is not possible since yaml files may refer to other yaml files, and in most cases they do. Even it would work (and it wouldn't) - it won't make reviewing code easier, since it will cause developers to run tlvf to generate code instead of just browsing it, which beats the purpose IMO.

RanRegev replied:
Fine.
Again - I'm not complaining on tlvf.
The question in this discussion is *why do we keep products in git?*
I see lots of drawbacks doing so, and I am not sure git is the correct solution to the problem raised when it was chosen. _I might be wrong_ by not understanding the problem from the first place.
Currently I don't work with tlvs, change them or have anything with them. When I will the _last_ thing that I'll do is to suggest a change in the mechanism. And that I will do *only after* I have tried everything with all the knowledge I can get.
Suggesting changing without understanding the problem is a _huge_ mistake. I'm not doing that.
The point here is not the tlv - but saving compilation (a.k.a generated) results.

tomereli reviewed (approved):
There is one major comment by Arnout which is not addressed. Other than that LGTM

rmelotte commented:
Rebased on master, fixed some of the nitpicks.
